tango/integrations/torch/train.py

```diff
@@ -460,25 +412,107 @@ def _train(
+            config.train_steps,
         )
     )
 
+    def is_best_checkpoint() -> bool:
+        """
+        A closure that we'll call when saving checkpoints to check if we should symlink
+        the best checkpoint path to the current checkpoint file.
+        """
+        if val_metric is not None and best_val_metric is not None:
+            return (config.minimize_val_metric and val_metric <= best_val_metric) or (
+                not config.minimize_val_metric and val_metric >= best_val_metric
+            )
+        elif best_batch_loss is not None:
+            return best_batch_loss <= batch_loss
+        else:
+            return False
+
+    def save_state(step: int):
+        """
+        A closure that we'll call every `checkpoint_every` steps in the train loop to
+        save model and training state.
+        """
+        temp_state_file = tempfile.NamedTemporaryFile(
+            "w+b", dir=config.work_dir, delete=False, suffix=".pt"
+        )
+        try:
+            with Tqdm.wrapattr(
+                temp_state_file,
+                "write",
+                desc="Saving checkpoint",
+                leave=False,
+                disable=not config.is_local_main_process,
+            ) as f:
+                checkpoint_state = {
+                    "optimizer": optimizer.state_dict(),  # type: ignore[attr-defined]
+                    "scheduler": None
+                    if lr_scheduler is None
+                    else lr_scheduler.state_dict(),  # type: ignore[attr-defined]
+                    "model": model.module.state_dict()  # type: ignore[attr-defined]
+                    if config.is_distributed
+                    else model.state_dict(),  # type: ignore[attr-defined]
+                    "training_steps": step + 1,
+                    "best_batch_loss": best_batch_loss,
+                    f"val_{config.val_metric_name}": val_metric,
+                    f"best_{config.val_metric_name}": best_val_metric,
+                    "callbacks": [
+                        callback.state_dict() for callback in callbacks  # type: ignore[union-attr]
+                    ],
+                }
+                for callback in callbacks:  # type: ignore[union-attr]
+                    callback.pre_checkpoint(checkpoint_state)  # type: ignore[union-attr]
+                torch.save(checkpoint_state, f)
+            temp_state_file.close()
+            os.replace(temp_state_file.name, config.state_path_for_step(step))
+
+            # Link to most recent state path.
+            if config.state_path.is_symlink():
+                config.state_path.unlink()
+            config.state_path.symlink_to(config.state_path_for_step(step))
+
+            # Link to best state path.
+            if is_best_checkpoint():
+                if config.best_state_path.is_symlink():
+                    config.best_state_path.unlink()
+                config.best_state_path.symlink_to(config.state_path_for_step(step))
```

Comment:
I thought the best state was saved as a hardlink, not a symlink, so it doesn't break when the file that's named after the number of steps goes away.

Reply:
Good point. Changed.